### PR TITLE
Fix issues around component classes / (text) opacity

### LINKED ISSUE #3413: Rich styles computation for dom nodes doesn't take text-opacity into account
`DOMNode.rich_style` does not take into account the text opacity when computing the foreground colour.

This will _likely_ fix [Textual#3342](https://github.com/Textualize/textual/issues/3342), which in turn will _likely_ fix [Textual#3304](https://github.com/Textualize/textual/issues/3304).

### PR DESCRIPTION
**Pull request by rodrigogiraoserrao**
This PR will fix #3413, #3342, and #3304.

This started out as a PR that intended to fix an issue with the option list and kept growing in size as I kept uncovering more issues with component classes & styles, (partial) rich styles, and interactions with (text-)opacity styles.

I've stopped digging because there's some things with component classes we'd like to iron out and those will probably help with what's not handled in this PR yet.

Some things that were left to be done/fixed/considered:
 - #3421 
 - #3467

### CODE REVIEW

**Comment by rodrigogiraoserrao** [comment]
de6a479 introduces a snapshot test app that you can run with `textual run tests/snapshot_tests/snapshot_apps/component_classes_opacity.py`.

The “empty boxes” are a good thing because I'm using component classes to set colours to transparent.
The screenshot shows that the only widgets that aren't fixed are `DirectoryTree`, `Tree`, `Checkbox`, and `RadioButton` (i.e., we need to fix `Tree` and `ToggleButton`).

Screenshot 2023-09-27 at 17 43 13

The screenshot shows that the previous commits fix #3304.

**Comment by rodrigogiraoserrao** [comment]
The reason the (directory) tree and toggle button labels are still white is because they use partial rich styles to enable markup to be used in the labels, which don't currently take into account text opacity.

To know how to resolve this, we need to know what should happen if a widget has mark up and conflicting TCSS.
For example, what should the checkbox `Checkbox("[red bold]This is just[/] some text.")` look like if the following TCSS is specified:

```sass
Checkbox > .toggle--label {
    color: blue;
    text-opacity: 50%;
}
```

1. The red/bold remains, but faded 50% and the “some text” is blue 50%
2. The red/bold remains and not faded, “some text” is blue 50%
3. The bold remains, the whole text is blue 50%
4. The TCSS overrides the markup completely and everything is just blue 50%
5. The existence of markup overrides the TCSS completely and the TCSS is essentially ignored
6. Some other combination I'm not considering 

After I have the answer to this question, I can get to fixing the code so that it complies with what we decide.

File: src/textual/widgets/_tabs.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ class Tab(Static):
     }
     Tab:disabled {
         color: $text-disabled;
-        text-opacity: 50%;
+        text-opacity: 70%;
```
Comment (rodrigogiraoserrao) [inline src/textual/widgets/_tabs.py:110]: This accounts for the now-correct calculation of the final text color when opacity is present.

**Comment by willmcgugan** [comment]
I have an alternative solution in mind for `_redirect_component_styles`. Solution is good, but I want to make this change as part of a larger update.

**Pull request closed without merging**